portable/NetworkInterface/LPC54018/NetworkInterface.c

```diff
@@ -129,197 +131,205 @@ void ENET_IntCallback( ENET_Type * base,
-
-    if( xTaskCreate( rx_task, "rx_task", 512, NULL, ( configMAX_PRIORITIES - 1 ), &receiveTaskHandle ) != pdPASS )
-    {
-        PRINTF( "Network Receive Task creation failed!.\n" );
-
-        while( 1 )
-        {
-        }
-    }
-
-    return pdTRUE;
+BaseType_t returnValue = pdFALSE;
+static enum {initPhy, startReceiver, waitForLink, configurePhy }networkInitialisePhase = initPhy;
+
+	switch(networkInitialisePhase)
+	{
+	default:
+		networkInitialisePhase = initPhy;
+	case initPhy:
+		{
+
+			phy_config_t phyConfig;
+			phyConfig.phyAddr = PHY_ADDRESS;
+			phyConfig.autoNeg = true;
+			mdioHandle.resource.base = ENET;
+
+			status_t status = PHY_Init( &phyHandle, &phyConfig );
+			if(status == kStatus_PHY_AutoNegotiateFail)
+			{
+				PRINTF( "\nPHY Auto-negotiation failed. Please check the cable connection and link partner setting.\n" );
+				break;
+			}
+		}
+	case startReceiver:
+		networkInitialisePhase = startReceiver;
+		if( xTaskCreate( rx_task, "rx_task", 512, NULL, ( configMAX_PRIORITIES - 1 ), &receiveTaskHandle ) != pdPASS )
+		{
+			PRINTF( "Network Receive Task creation failed!.\n" );
+			break;
+		}
+	case waitForLink:
+		networkInitialisePhase = waitForLink;
+		if(!xGetPhyLinkStatus())
+		{
+			PRINTF("No Link\n");
+			break;
+		}
+	case configurePhy:
+		{
+			networkInitialisePhase = configurePhy;
+			enet_config_t config;
+			phy_speed_t speed;
+			phy_duplex_t duplex;
+			PHY_GetLinkSpeedDuplex( &phyHandle, &speed, &duplex );
+			/* Get default configuration 100M RMII. */
+			ENET_GetDefaultConfig( &config );
+
+			/* Use the actual speed and duplex when phy success to finish the autonegotiation. */
+			config.miiSpeed = ( enet_mii_speed_t ) speed;
+			config.miiDuplex = ( enet_mii_duplex_t ) duplex;
+
+			/* Initialize ENET. */
+			uint32_t refClock = 50000000;     /* 50MHZ for rmii reference clock. */
+			ENET_Init( ENET, &config, g_macAddr, refClock );
+
+			/* Enable the rx interrupt. */
+			ENET_EnableInterrupts( ENET, ( kENET_DmaRx ) );
+
+			/* Initialize Descriptor. */
+			int bufferIndex;
+			for( bufferIndex = 0; bufferIndex < ENET_RXBD_NUM; bufferIndex++ )
+			{
+				rxbuffer[ bufferIndex ] = ( uint32_t ) &receiveBuffer[ bufferIndex ];
+			}
+			/* prepare the buffer configuration. */
+			enet_buffer_config_t buffConfig[ 1 ] =
+			{
+				{
+					ENET_RXBD_NUM, ENET_TXBD_NUM,
+					&g_txBuffDescrip[ 0 ], &g_txBuffDescrip[ 0 ],
+					&g_rxBuffDescrip[ 0 ], &g_rxBuffDescrip[ ENET_RXBD_NUM ],
+					&rxbuffer[ 0 ], ENET_BuffSizeAlign( ENET_RXBUFF_SIZE ),
+				}
+			};
+			ENET_DescriptorInit( ENET, &config, &buffConfig[ 0 ] );
+
+			/* Create the handler. */
+			ENET_CreateHandler( ENET, &g_handle, &config, &buffConfig[ 0 ], ENET_IntCallback, NULL );
+			NVIC_SetPriority( 65 - 16, 4 ); /* TODO this is a hack and I would expect a nice ENET API for priority. */
+
+			/* Active TX/RX. */
+			ENET_StartRxTx( ENET, 1, 1 );
+
+			networkInitialisePhase = initPhy;
+			returnValue = pdTRUE;
+		}
```

Comment:
That is a good solution: a switch that handles all initialisation in stages!

I do understand why you put the status to `initPhy` again? The task has been created, you don't want to do that again.

It is preferred to let `xNetworkInterfaceInitialise()` return pdPASS and pdFAIL, rather then pdTRUE and pdFALSE.

Also, I would call `xTaskCreate()` in the latest stage because the RX-talk will take over the access to the PHY. Now there might be concurrent access.

Reply:
Moved xTaskCreate to the end.  Also switched for pdPASS/pdFAIL

The last initPhy was out of concern for reinitialising the network.  I removed it so you cannot reinitialise the network.  If this is needed we will need to think about it.